docs/developer/designs/topology-awareness/README.md

```diff
@@ -0,0 +1,287 @@
+2. For each possible topology hierarchy that has enough resources:
+   a. Add a virtual node selector to the job and simulate allocation
+
+* This will work best if the first approach is implemented first
+* Significant performance reduction in big clusters is expected here.
+
+
+### Topology Info struct
+
+Both approaches need to build a topology tree and update it with the changes the scheduler is doing during the session and simulation.
+We will read the Topology resources during the cluster session start phase and create TopologyInfo structs that will be maintained throughout the session with every change.
+
+#### Suggested TopologyInfo Structure
+
+```go
+// TopologyDomainID uniquely identifies a topology domain
+type TopologyDomainID string
+
+// TopologyInfo represents a topology tree for the cluster
+type TopologyInfo struct {
+    // Root of the topology tree
+    Root *TopologyDomainInfo
+
+    // Map of all domains by their ID for quick lookup
+    Domains map[TopologyDomainID]*TopologyDomainInfo
+
+    // Name of this topology configuration
+    Name string
+}
+
+// TopologyDomainInfo represents a node in the topology tree
+type TopologyDomainInfo struct {
+    // Unique ID of this domain
+    ID TopologyDomainID
+
+    // Name of this domain
+    Name string
+
+    // Level in the hierarchy (e.g., "datacenter", "zone", "rack", "node")
+    Level string
+
+    // Parent domain, nil for root
+    Parent *TopologyDomainInfo
+
+    // Child domains
+    Children []*TopologyDomainInfo
+
+    // Nodes that belong to this domain
+    Nodes map[string]*node_info.NodeInfo
+
+    // Total available resources in this domain
+    AvailableResources v1.ResourceList
+
+    // Total allocated resources in this domain
+    AllocatedResources v1.ResourceList
+
+    // Depth in the tree from root (0 for root)
+    Depth int
+}
+```
+
+There are two possibilities as to where the struct should be built and kept:
+  * Plugin - We will use a plugin that will build it in `OnSessionOpen` and will use hooks similar to DRA plugin to keep tabs on all the changes made in the cluster.
+  * ClusterInfo - This can be done with TopologyInfo structs in the ClusterInfo package, the plugin will later read this from the cache to make node order decision and will update this struct with simulated allocations.
+Because this might be resource intensive we will leave users with the option to totally disable it and only reference it in the plugin.
+
+
+
+### Topology constraint request on a job
+
+The pod-grouper will look for topology annotations on the top owner of pods and set it to fields on the PodGroup that it creates.
+
+#### Annotations for Topology Constraints
+
+To support topology-aware scheduling, we'll define the following annotations that can be added to job resources (e.g., Job, BatchJob, MPIJob):
+
+1. `scheduling.kai.nvidia.com/topology-constraint-type`: Specifies whether the topology constraint is "Required" or "Preferred"
```

Comment:
maybe `kai.scheduler` so it will be consistent with other annotation prefixes?

Reply:
:+1: I mentioned somewhere that I want to fix it, but not yet because I don't want the change to hide other conversations.